### PR TITLE
RavenDB-17133 - Storage Report times out if database has attachments

### PR DESCRIPTION
**Pull request by shiranshalom**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17133

### Type of change

- Bug fix
- Optimization

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- Tests have been added that prove the fix is effective or that the feature works
- No documentation update is needed 

### Testing 

- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- It requires further work in the Studio

### CODE REVIEW

File: src/Voron/Debugging/StorageReportGenerator.cs [hunk 1]
```diff
@@ -322,8 +322,34 @@ public static TreeReport GetReport(Tree tree, bool includeDetails)
...
+                return new StreamsReport
+                {
+                    AllocatedSpaceInBytes = -1,
+                    NumberOfStreams = tree.State.NumberOfEntries,
+                    Streams = new List<StreamDetails>
+                    {
+                        new StreamDetails
+                        {
+                            Name = "Stream details summary info",
+                            AllocatedSpaceInBytes = 1024,
```
Comment (ppekrol) [inline src/Voron/Debugging/StorageReportGenerator.cs]: why 1024?

File: src/Voron/Debugging/StorageReportGenerator.cs [hunk 2]
```diff
@@ -322,8 +322,34 @@ public static TreeReport GetReport(Tree tree, bool includeDetails)
...
+            {
+                // there are cases we don't need/want to pull the entire data for the report
+                // so we create a report that includes the metadata info
+                // without loading the actual data 
+
+                return new StreamsReport
+                {
+                    AllocatedSpaceInBytes = -1,
+                    NumberOfStreams = tree.State.NumberOfEntries,
+                    Streams = new List<StreamDetails>
```
Comment (ppekrol) [inline src/Voron/Debugging/StorageReportGenerator.cs:387]: the item inside look fake, do we really need it?

Comment (ayende) [reply src/Voron/Debugging/StorageReportGenerator.cs:387]: It is fake, we need that so the studio will show the entry, because without the details, we are basically lying to the user.

Something to check, we may want to actually report this in inverse. So we figure out how much total space we *aren't* accounting for, and assign it to this value.

**Comment by arekpalinski** [comment]
@shiranshalom Please add a test for that. See `StorageReportGenerationTests.TreeReportContainsInfoAboutStreams()`

File: src/Voron/Debugging/StorageReportGenerator.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Voron/Debugging/StorageReportGenerator.cs:387]: I think we still need to go over the streams that we have there and calculate `StreamsReport.TotalNumberOfAllocatedPages`. I think that just calculating the total size is fine.  This way we'll be able to properly display the size of `Trees`:

image


and  `Attachments` tree here:

image


What we _don't_ need to include are the details of each stream (`StreamDetails`). We need those _only_ if a user goes deeper to see that:

image

**Comment by ayende** [comment]
What is the status here?

File: src/Voron/Debugging/StorageReportGenerator.cs [hunk 3]
```diff
@@ -90,26 +91,29 @@ public DetailedStorageReport Generate(DetailedReportInput input)
...
             var trees = new List<TreeReport>();
-
             foreach (var tree in input.Trees)
             {
                 var treeReport = GetReport(tree, input.IncludeDetails);
-
                 trees.Add(treeReport);
             }
 
+            long fixedTreesAllocatedSpaceInBytes = 0;
```
Comment (ayende) [inline src/Voron/Debugging/StorageReportGenerator.cs]:
```suggestion
            long _fixedTreesAllocatedSpaceInBytes = 0;
```

File: src/Voron/Debugging/StorageReportGenerator.cs [hunk 4]
```diff
@@ -90,26 +91,29 @@ public DetailedStorageReport Generate(DetailedReportInput input)
...
             foreach (var fst in input.FixedSizeTrees)
             {
                 var treeReport = GetReport(fst, input.IncludeDetails);
-
                 trees.Add(treeReport);
+
+                fixedTreesAllocatedSpaceInBytes += treeReport.AllocatedSpaceInBytes;
             }
 
+            long tablesAllocatedSpaceInBytes = 0;
```
Comment (ayende) [inline src/Voron/Debugging/StorageReportGenerator.cs]:
```suggestion
            long _tablesAllocatedSpaceInBytes = 0;
```

File: src/Voron/Debugging/StorageReportGenerator.cs (hunk 3 above)
Comment (shiranshalom) [reply src/Voron/Debugging/StorageReportGenerator.cs]: done

File: src/Voron/Debugging/StorageReportGenerator.cs (hunk 4 above)
Comment (shiranshalom) [reply src/Voron/Debugging/StorageReportGenerator.cs]: done

File: src/Voron/Debugging/StorageReportGenerator.cs [hunk 5]
```diff
@@ -121,15 +125,37 @@ public DetailedStorageReport Generate(DetailedReportInput input)
...
+                // so we calculate the original size as if we read the streams by:
+                // [DataFile allocated space] - [DataFile free space] - [Tables allocated space] - [FixedTrees allocated space] - [pre allocated buffers space] 
+
+                var skippedStreamsAllocatedSpaceInBytes = dataFile.AllocatedSpaceInBytes - dataFile.FreeSpaceInBytes - _tablesAllocatedSpaceInBytes  - preAllocatedBuffers.OriginallyAllocatedSpaceInBytes - _fixedTreesAllocatedSpaceInBytes ;
+                _skippedStreamsDetailsEntry.AllocatedSpaceInBytes = skippedStreamsAllocatedSpaceInBytes;
+                _skippedStreamsDetailsEntry.Length = skippedStreamsAllocatedSpaceInBytes;
+
+                foreach (var tree in trees)
+                {
+                    if (tree.Streams?.Streams != null && tree.Streams.Streams[0].Name == "Stream details summary info")
```
Comment (arekpalinski) [inline src/Voron/Debugging/StorageReportGenerator.cs]: This ""Stream details summary info"" should go to a const and be used in both places

File: test/SlowTests/Voron/Issues/RDBCL_1977.cs [hunk 6]
```diff
@@ -4,93 +4,95 @@
 using Xunit;
 using Xunit.Abstractions;
 
-namespace SlowTests.Voron.Issues;
-
-public class RDBCL_1977 : StorageTest
+namespace SlowTests.Voron.Issues
```
Comment (arekpalinski) [inline test/SlowTests/Voron/Issues/RDBCL_1977.cs]: Please revert the changes made here. I already did it on 4.2 branch so it will be merged to other branches

https://github.com/ravendb/ravendb/pull/13086

Comment (arekpalinski) [reply test/SlowTests/Voron/Issues/RDBCL_1977.cs]: Note: our CI environments already have .net sdk 6.0 installed so it will compile correctly

Comment (shiranshalom) [reply test/SlowTests/Voron/Issues/RDBCL_1977.cs]: done

File: src/Voron/Debugging/StorageReportGenerator.cs (hunk 5 above)
Comment (shiranshalom) [reply src/Voron/Debugging/StorageReportGenerator.cs]: done

File: src/Voron/Constants.cs [hunk 7]
```diff
@@ -91,7 +91,9 @@ static Tree()
         public static readonly Slice DatabaseFilenameSlice;
         public const string DatabaseRecordFileNameInBackup = "___database_record___";
 
-        public const int DefaultMaxLogLengthBeforeCompaction = 64; //how much entries in log to keep before compacting it into snapshot       
+        public const int DefaultMaxLogLengthBeforeCompaction = 64; //how much entries in log to keep before compacting it into snapshot
+
+        public const string SkippedStreamsDetailsName = "Stream details summary info";
```
Comment (arekpalinski) [inline src/Voron/Constants.cs]: In this file we have constants that have fundamental / global meaning for the Voron itself. The one that you added is very specific to the storage report generation - which isn't a fundamental functionality.

I think a way better candidate would be to put it in `StorageReportGenerator`. Especially that we use it only there

File: src/Voron/Debugging/StorageReportGenerator.cs [hunk 8]
```diff
@@ -121,15 +125,37 @@ public DetailedStorageReport Generate(DetailedReportInput input)
...
+                // so we calculate the original size as if we read the streams by:
+                // [DataFile allocated space] - [DataFile free space] - [Tables allocated space] - [FixedTrees allocated space] - [pre allocated buffers space] 
+
+                var skippedStreamsAllocatedSpaceInBytes = dataFile.AllocatedSpaceInBytes - dataFile.FreeSpaceInBytes - _tablesAllocatedSpaceInBytes  - preAllocatedBuffers.OriginallyAllocatedSpaceInBytes - _fixedTreesAllocatedSpaceInBytes ;
+                _skippedStreamsDetailsEntry.AllocatedSpaceInBytes = skippedStreamsAllocatedSpaceInBytes;
+                _skippedStreamsDetailsEntry.Length = skippedStreamsAllocatedSpaceInBytes;
+
+                foreach (var tree in trees)
+                {
+                    if (tree.Streams?.Streams != null && tree.Streams.Streams[0].Name == Constants.SkippedStreamsDetailsName)
```
Comment (arekpalinski) [inline src/Voron/Debugging/StorageReportGenerator.cs]: Just in case - let's check that `tree.Streams.Streams.Count > 0`

**Comment by arekpalinski** [comment]
@ppekrol this PR isn't going to be merged until @Danielle9897 handles the studio part and push it here (there is dedicated issue for that https://issues.hibernatingrhinos.com/issue/RavenDB-17465)

File: src/Voron/Debugging/StorageReportGenerator.cs (hunk 8 above)
Comment (shiranshalom) [reply src/Voron/Debugging/StorageReportGenerator.cs]: done

File: src/Voron/Constants.cs (hunk 7 above)
Comment (shiranshalom) [reply src/Voron/Constants.cs]: done

**Review by arekpalinski — APPROVED** [review]

**Comment by arekpalinski** [comment]
Status update. @ml054 found some issue after integrating that with the studio, @shiranshalom will fix it

File: test/SlowTests/Voron/Storage/StorageReportGenerationTests.cs [hunk 9]
```diff
@@ -390,6 +391,54 @@ public void TreeReportContainsInfoAboutStreams(int numberOfStreams, int seed)
...
+                var treeReport = report.Trees.Find(x => x.Name == "streams-tree");
+                var fullTreeReport = fullReport.Trees.Find(x => x.Name == "streams-tree");
+                
+                Assert.NotNull(treeReport);
+                Assert.NotNull(fullTreeReport);
+
+                Assert.Equal(1, treeReport.Streams.Streams.Count);
+                Assert.Equal(StorageReportGenerator.SkippedStreamsDetailsName, treeReport?.Streams.Streams[0].Name);
+
+                var streamsSizeInMb = Math.Ceiling(treeReport.Streams.Streams[0].AllocatedSpaceInBytes * 0.000001);
```
Comment (ml054) [inline test/SlowTests/Voron/Storage/StorageReportGenerationTests.cs:435]: ?

Comment (shiranshalom) [reply test/SlowTests/Voron/Storage/StorageReportGenerationTests.cs:435]: In `includeDetails: false` case, at the beginning, we are not reading the streams -> we don't have the actual size of each stream. 
For each tree report which contains streams -> we are calculating its size [roughly] by using other parameters we do have - but there may be cases we can miss something on calculation. 
Thus, for skipped streams/trees we can get a difference in sizes [really small one]. For example, we can get 8KB for the calculated ["fake"] value instead of 7.99KB for the original one.

**Review by arekpalinski — APPROVED** [review]

**Comment by arekpalinski** [comment]
It's ready to merge

**Comment by arekpalinski** [comment]
@ml054 is there anything else you need to handle here after @shiranshalom's changes?

**Comment by ml054** [comment]
> @ml054 is there anything else you need to handle here after @shiranshalom's changes?

No

**Comment by shiranshalom** [comment]
It's ready to merge
